gqlalchemy/instance_runner.py

```diff
@@ -0,0 +1,228 @@
+# Copyright (c) 2016-2022 Memgraph Ltd. [https://memgraph.com]
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import docker
+import os
+import psutil
+import socket
+import subprocess
+import time
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Any, Dict, Optional, Union
+from .memgraph import Memgraph
+
+
+class DockerImage(Enum):
+    MEMGRAPH = "memgraph/memgraph"
+    MAGE = "memgraph/memgraph-mage"
+
+
+def wait_for_port(host: str = "127.0.0.1", port: int = 7687, delay: float = 0.01, timeout: float = 5.0) -> None:
+    """Wait for a TCP port to become available.
+
+    Args:
+        host: A string representing the IP address that is being checked.
+        port: A string representing the port that is being checked.
+        delay: A float that defines how long to wait between retries.
+        timeout: A float that defines how long to wait for the port.
+
+    Raises:
+      TimeoutError: Raises an error when the host and port are not accepting
+        connections after the timeout period has passed.
+    """
+    start_time = time.perf_counter()
+    time.sleep(delay)
+    while True:
+        try:
+            with socket.create_connection((host, port), timeout=timeout):
+                break
+        except OSError as ex:
+            time.sleep(delay)
```

Comment:
We sleep 10ms every time? Is that a lot of pinging to Memgraph connection?
What do you think @gitbuda

Reply:
Since delay is configurable, I think it's fine, but ofc, exponential backoff would be better here 😄